packages/core/prompts/openai.tsx

```diff
@@ -0,0 +1,241 @@
+                  arguments: function_call.arguments!,
+                },
+        } satisfies ChatCompletionMessageParam;
+      }
+
+      case 'function': {
+        if (message.name === undefined) {
+          throw new Error('Invalid function call in message. Expected a name');
+        }
+
+        return {
+          role: message.role,
+          content: message.content,
+          name: message.name,
+        } satisfies ChatCompletionMessageParam;
+      }
+    }
+  });
+}
+
+// copy of open ai messages (so we don't have a dependency on the openai package)
+export type ChatCompletionMessageParam =
+  | ChatCompletionSystemMessageParam
+  | ChatCompletionUserMessageParam
+  | ChatCompletionAssistantMessageParam
+  | ChatCompletionToolMessageParam
+  | ChatCompletionFunctionMessageParam;
+
+export interface ChatCompletionSystemMessageParam {
+  /**
+   * The contents of the system message.
+   */
+  content: string | null;
+
+  /**
+   * The role of the messages author, in this case `system`.
+   */
+  role: 'system';
+}
+
+export interface ChatCompletionUserMessageParam {
+  /**
+   * The contents of the user message.
+   */
+  content: string | Array<ChatCompletionContentPart> | null;
+
+  /**
+   * The role of the messages author, in this case `user`.
+   */
+  role: 'user';
+}
+
+export type ChatCompletionContentPart =
+  | ChatCompletionContentPartText
+  | ChatCompletionContentPartImage;
+
+export interface ChatCompletionContentPartText {
+  /**
+   * The text content.
+   */
+  text: string;
+
+  /**
+   * The type of the content part.
+   */
+  type: 'text';
+}
+
+export interface ChatCompletionContentPartImage {
+  image_url: ChatCompletionContentPartImage.ImageURL;
+
+  /**
+   * The type of the content part.
+   */
+  type: 'image_url';
+}
+
+export namespace ChatCompletionContentPartImage {
+  export interface ImageURL {
+    /**
+     * Specifies the detail level of the image.
+     */
+    detail?: 'auto' | 'low' | 'high';
+
+    /**
+     * Either a URL of the image or the base64 encoded image data.
+     */
+    url?: string;
+  }
+}
+
+export interface ChatCompletionAssistantMessageParam {
+  /**
+   * The contents of the assistant message.
+   */
+  content: string | null;
+
+  /**
+   * The role of the messages author, in this case `assistant`.
+   */
+  role: 'assistant';
+
+  /**
+   * Deprecated and replaced by `tool_calls`. The name and arguments of a function
+   * that should be called, as generated by the model.
+   */
+  function_call?: ChatCompletionAssistantMessageParam.FunctionCall;
+
+  /**
+   * The tool calls generated by the model, such as function calls.
+   */
+  tool_calls?: Array<ChatCompletionMessageToolCall>;
+}
+
+export namespace ChatCompletionAssistantMessageParam {
+  /**
+   * Deprecated and replaced by `tool_calls`. The name and arguments of a function
+   * that should be called, as generated by the model.
+   */
+  export interface FunctionCall {
```

Comment:
Some of these are duplicates of `core/shared/types.ts`. I'd prefer not maintaining two copies of them, maybe these should be moved there, or a new openai types file?

Reply:
A new `openai` types file sounds good. I would ideally want to clearly separate the AI SDK message types from the OpenAI message types to prevent confusion and avoid errors in future changes (when they might evolve indepedently).

Reply:
@MaxLeiter actually it looks like it's only `FunctionCall` that's fully overlapping, and I'd argue that the use in `types.ts` is as part of AI SDK messages, and here it is used as an OpenAI message. The evolution forces for both of them are different (here: OpenAI changes their API, in `types.ts`: we decide to change AI SDK messages), so I'd rather keep them separate and remove the comment in `types.ts`. What do you think?

Reply:
Yeah, this makes sense. We'll need to tackle the type story more definitively in the future, but for now let's ship both and keep OpenAI vs AI SDK separate.